### PR TITLE
Redirect users to https when logging in and accessing the admin

### PR DESCRIPTION
**Pull request by nathanbegbie**
Ticket: https://praekeltorg.atlassian.net/browse/UNICEF-298

This PR will add middleware that will redirect users to use `https` when logging in, registering or accessing the admin.

[Link 1](https://gist.github.com/bryanchow/1035190/c0b060ccee057aa803cfa586d845b2049bb3560a)
[Link 2](https://bitbucket.org/nextscreenlabs/django-ssl-redirect/src/95da4b1eab0ad00a00bcb7b9fb09f02fa8ee7019/ssl_redirect/middleware.py?at=default&fileviewer=file-view-default)

### CODE REVIEW

**Comment by miltontony** [comment]
👌 looks good from me. @JayH5 will rely on you for a final 👍 since you discussed the solution with the team.

File: iogt/middleware.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        request_is_secure = self.request_is_secure(request)
+        if response_should_be_secure and not request_is_secure:
+            return HttpResponsePermanentRedirect(
+                "https://{}{}".format(request.get_host(),
+                                      request.get_full_path())
+            )
+
+    def response_should_be_secure(self, request, HTTPS_PATHS):
+        for path in HTTPS_PATHS:
+            if request.path.startswith(u'/{}'.format(path)):
```
Comment (JayH5) [inline iogt/middleware.py]: Is `request.path` a string (I don't know if there's some Django magic here)? If it is, `startswith` is not going to necessarily match parent paths only, e.g. both `/admin` and `/admin-django` could be matched by `/admin` which may be unexpected. The solution would probably be to parse the path into some kind of proper URL abstraction.

Also, personally I'd just require the `HTTPS_PATHS` setting values to start with `/`--or this would be another thing a URL library would take care of. If somebody _did_ put a path in the settings with a leading `/` it wouldn't work as expected currently.

_Also_, tiny nit... having an uppercase parameter name is weird to me 😅.

File: iogt/tests/test_middleware.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        request = self.factory.get('/', **headers)
+
+        middleware = SSLRedirectMiddleware()
+        response = middleware.process_request(request)
+
+        self.assertEqual(response, None)
+
+    def test_no_redirect_when_secure(self):
+        headers = {'HTTP_X_FORWARDED_PROTO': 'https'}
+        request = self.factory.get('/admin/', **headers)
```
Comment (JayH5) [inline iogt/tests/test_middleware.py:101]: Edge case with absent/present trailing `/` should probably be tested.

**Review by JayH5 — COMMENTED** [review]
Sorry for being nitpicky, but if this is a security-critical thing we need to make sure it's fool-proof.

File: iogt/tests/test_middleware.py (hunk 2 above)
Comment (nathanbegbie) [reply iogt/tests/test_middleware.py:101]: The use of the new `clean_path` ensure that URL paths are represented as a list of segments - so we can rest assured that no matter how it's declared in the settings file, it will be represented consistently in the middleware
e.g. `/admin/login` -> `['admin', 'login']`
`admin/login` -> `['admin', 'login']`
`admin/login/` -> `['admin', 'login']`

File: iogt/middleware.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def clean_path(path):
+    '''
+    Converts string of URL paths to list of path elements
+    '''
+    if path == u'/':
+        return [u'/']
+    else:
+        return filter(lambda a: a != '', path.split('/'))
```
Comment (JayH5) [inline iogt/middleware.py]: Nothing _technically_ wrong with paths with multiple `/`. We just want to ignore a first leading or final trailing one. But they're also rare/weird enough to maybe not matter.

`filter` will probably make this Python 3-unfriendly.

URLs are hard I'm sorry :(

Comment (nathanbegbie) [reply iogt/middleware.py]: @JayH5 I can update `filter` to `list(filter(...)` to make it python3 compatible.
But do you want me to revert back to checking paths as strings as opposed to lists, assuming I've checked for trailing and leading slashes?

Comment (JayH5) [reply iogt/middleware.py]: @nathanbegbie list comp is probably best:
```python
return [a for a in path.split(u'/') if a]
```
Still think comparing segments is definitely better than strings.

**Comment by nathanbegbie** [comment]
@JayH5 ready for re-review

**Review by JayH5 — APPROVED** [review]
:+1:
